src/ips.rs

```diff
@@ -0,0 +1,445 @@
+
+        self.eigenvector_factors = NormalizationFactors {
+            min,
+            max,
+            value: NORMALIZE_TO_VALUE,
+        };
+    }
+
+    fn determine_betweenness_factors(&mut self, nodes: &[Node]) {
+        let min = nodes
+            .iter()
+            .min_by(|a, b| a.betweenness.partial_cmp(&b.betweenness).unwrap())
+            .map(|m| m.betweenness)
+            .unwrap_or(0.0);
+        let max = nodes
+            .iter()
+            .max_by(|a, b| a.betweenness.partial_cmp(&b.betweenness).unwrap())
+            .map(|m| m.betweenness)
+            .unwrap_or(NORMALIZE_TO_VALUE);
+
+        self.betweenness_factors = NormalizationFactors {
+            min,
+            max,
+            value: NORMALIZE_TO_VALUE,
+        };
+    }
+
+    fn determine_closeness_factors(&mut self, nodes: &[Node]) {
+        let min = nodes
+            .iter()
+            .min_by(|a, b| a.closeness.partial_cmp(&b.closeness).unwrap())
+            .map(|m| m.closeness)
+            .unwrap_or(0.0);
+        let max = nodes
+            .iter()
+            .max_by(|a, b| a.closeness.partial_cmp(&b.closeness).unwrap())
+            .map(|m| m.closeness)
+            .unwrap_or(NORMALIZE_TO_VALUE);
+
+        self.closeness_factors = NormalizationFactors {
+            min,
+            max,
+            value: NORMALIZE_TO_VALUE,
+        };
+    }
+
+    fn rate_node(&self, node: &Node, degree: u32, eigenvalue: f64) -> f64 {
+        // Calculate rating for node
+
+        // Rating is a combination of the following factors:
+        // 1. Degree
+        let mut rating = ((degree as f64 - self.degree_factors.min)
+            / (self.degree_factors.max - self.degree_factors.min)
+            * self.degree_factors.value)
+            * self.config.degree_weight;
+        // 2. Betweenness
+        rating += ((node.betweenness - self.betweenness_factors.min)
+            / (self.betweenness_factors.max - self.betweenness_factors.min)
+            * self.betweenness_factors.value)
+            * self.config.betweenness_weight;
+        // 3. Closeness
+        rating += ((node.closeness - self.closeness_factors.min)
+            / (self.closeness_factors.max - self.closeness_factors.min)
+            * self.closeness_factors.value)
+            * self.config.closeness_weight;
+        // 4. Eigenvector
+        rating += ((eigenvalue - self.eigenvector_factors.min)
+            / (self.eigenvector_factors.max - self.eigenvector_factors.min)
+            * self.eigenvector_factors.value)
+            * self.config.eigenvector_weight;
+
+        rating
+    }
+
+    // Very simple algorithm to detect islands.
+    // Take first vertex and do BFS to find all connected vertices. If there are any unvisited vertices
+    // create new island and do BFS one more time. Repeat until all vertices are visited.
+    fn detect_islands(&self, nodes: &[Node], agraph: &AGraph) -> Vec<HashSet<usize>> {
```

Comment:
Just something to think about...another todo item.  This should maybe be done in the `compute_betweenness_and_closeness`.  An enumeration of the islands (how many, and each node has an island id (index))

Reply:
We use the `nodes` argument to get its length? Could we remove it, or we could add a TODO comment here to indicate we'll use it for something else later 🕵️‍♀️

And it feels like the function asserts:
```rust
assert_eq!(nodes.len(), agraph.len());
```
Could we add it on top if we need `nodes` for something else later?

Reply:
You're right. agraph is sufficent for now (and probably it should be also after moving that function to spectre).
Changed.